Sources/GRDBQuery/Query.swift

```diff
@@ -405,9 +378,13 @@ public struct Query<Request: Queryable>: DynamicProperty {
         private var trackedRequest: Request?
         private var cancellable: AnyCancellable?
         
-        func update(configuration queryConfiguration: Configuration, database: Request.DatabaseContext) {
-            // Give up if the request is already tracked.
-            guard isAutoupdating else {
+        func update(
+            queryObservationEnabled: Bool,
+            configuration queryConfiguration: Configuration,
+            database: Request.DatabaseContext)
+        {
+            // Give up if observation is disabled
+            guard queryObservationEnabled else {
```

Comment:
I think it might be a good idea changing this to `queryObservationEnabled || value == nil`, so that the `View` always fetches an initial value on first `update()`.

The difference in `onAppear` would be that the previously fetched value is displayed instead of `defaultValue` as a placeholder when the query observation gets enabled.

Reply:
> I think it might be a good idea changing this to queryObservationEnabled || value == nil, so that the View always fetches an initial value on first update().

To get this behavior, you use `.onRender`.

Reply:
I should rename "initial rendering" to "initial body rendering" to make things more clear.

Reply:
From my understanding of the code `.onRender` will keep updating in the background until the first `onDisappear`.

Reply:
If I understand well, you'd like one single db access on the first `body` render, and then nothing until onAppear. Do I get this right?

Reply:
This means that you expect a long delay between the initial `body` render, and the first `onAppear`. In my own experiments, I could not quite witness such a behavior. Do you have a reproducing setup?

Reply:
```swift
TabView {
  ViewA().tabItem { Text("A") }

  ViewB().tabItem { Text("B") }
}
```

`update()` is called for `ViewA` and `ViewB` immediately, no matter which one is shown initially.
`onAppear()` is then only called after the respective tab is selected.

Reply:
OK, I don't know what I was looking at when performing my own tests.

So yes, definitely, in a TabView, `queryObservation(.onRender)` starts observing long before `.onAppear`.

What exactly isn't quite ok for you with `queryObservation(.onAppear)`? Any undesired side effect?

Reply:
For one of my tabs I have a fetch request that takes ~0.8 seconds.
With `.onAppear` it shows a blank view (the `defaultValue`) and only a second later the data appears.

Reply:
### With "prefetching"

https://user-images.githubusercontent.com/7303830/164888694-c0d31db4-da7e-4529-bf4d-2414340b5457.mov

### Without "prefetching"

https://user-images.githubusercontent.com/7303830/164888695-9c2ebb6b-e505-4cff-9e8a-972a97959fc4.mov

Reply:
In the current state of the PR, "prefetching" is achieved with `.onRender`.

I'm not sure what you want to achieve with your request for "prefetch, but not really" request (fetch on first update, then start observing onAppear). I do not see how displaying a blank view is worse than displaying obsolete data.

Reply:
I was under the impression that displaying obsolete data in exchange of freeing up resources by not observing in the background is the point of `queryObservationEnabled`.
The same happens when using `.onAppear` for example: we stop observing between `onDisappear` -> `onAppear`, so the data is obsolete in `onAppear`.

The problem I see with `.onRender` right now is that it initially observes the database in the background, but after the first `onAppear` -> `onDisappear` dance it then stops observing in the background.

Reply:
Let's try to list **desirable behaviors** that developers may look for:

- **No blank screen**: The user never sees a blank state (the request's default value).

- **No initial animation**: If the view animates database changes, the first rendering of database values is not animated.

- **Best effort refresh**: The probability that the user sees obsolete database values is as low as the observed database request is fast.

- **Pause observation when not visible**: Spare resources by stopping observation in the `onAppear/onDisappear` dance.

- **Does not observe until first appearance**: Spare resources by not observing until the view appears (useful in a `TabView`).

In the current state of the PR, you can not check all the boxes:

1. When the request publisher publishes its first value right on subscription (ValueObservation with `.immediate` scheduling)

    | | No blank screen | No initial animation | Best effort refresh | Pause observation when not visible | Does not observe until first appearance |
    |:-|:-:|:-:|:-:|:-:|:-:|
    | `.always`   | X | X | X |   |   |
    | `.onRender` | X | X | X | X |   |
    | `.onAppear` | X | ¹ | X | X | X |

    ¹ I could witness variable SwiftUI behaviors here.

2. When the request publisher does not publish its first value right on subscription (ValueObservation without `.immediate` scheduling).

    | | No blank screen | No initial animation | Best effort refresh | Pause observation when not visible | Does not observe until first appearance |
    |:-|:-:|:-:|:-:|:-:|:-:|
    | `.always`   |   |   | X |   |   |
    | `.onRender` |   |   | ¹ | X |   |
    | `.onAppear` |   |   | ¹ | X | X |

    ¹ Since observation was stopped `onDisappear`, the view always displays obsolete database values when it re-appears.

In your case, it looks like you would like `.onAppear` with an `.immediate` ValueObservation.

If you do not use `.immediate` for some reason, then you're left with the second table, with *very few* Xs... But I'm sure you see why asynchronous fetches kill almost all expectations.

> I think it might be a good idea changing this to `queryObservationEnabled || value == nil`, so that the View always fetches an initial value on first `update()`.

I consider this idea with interest, but it would not change the fact that the view would always displays obsolete database values when it re-appears. Your enhance your demo **for the first appearance only**. This does not look like a solid promise to me. Please change my mind!

As mentioned in the OP, you can perform your customized observation handling with the `\.queryObservationEnabled` environment value. Did you try implementing some custom logic with it?

By the way, thank you for this conversation. We clarify the behaviors, and we may find better names for `QueryObservation` cases in the process.

Reply:
@kabouzeid I had to perform multiple edits on the above comment. Now I think it is settled 😅

Reply:
> [...] the view would always display obsolete database values when it re-appears, with a user-visible update soon later.

@kabouzeid, If you are willing to pay this price, then maybe this means that you do not expect the database to change often. Maybe you even know that it will not change at all when your view is off-screen, because your view is the only place where the database can be modified (I don't know, but I make hypothesis compatible with your request).

In such a setup, you will not have a user interface that displays obsolete database values when the user changes tab, and updates database values soon later right under the user's nose.

The default @Query behavior (`.always`), as well as `.onRender`, look like they fit your need: both will prefetch and then remain idle until the database changes.

Reply:
I may have to tweak the names of the enum cases and improve the documentation, based on our conversation, but in my current understanding the PR in its current state matches a great deal of use cases, and is mergeable.

Reply:
Sorry for my delayed answer, I usually take the day off on Sundays.

Thanks for the overview with the tables, I mostly agree, though I could not reproduce `.onRender` and `.onAppear` behaving differently regarding "No blank screen" and "No initial animation" when using `.immediate`.

---

Indeed I was using
```swift
.publisher(in: database.databaseReader, scheduling: .async(onQueue: .global(qos: .background)))
```
because `.immediate` was not feasible with ~1sec fetch time, and even `.async(.main)` was not feasible when they were bursts of changes in the database.

I fixed my performance problems now by having a `Query` with `selectID()` for the `List` and then separate `Query`s for each cell, which observe the actual contents of the particular cells. The cell `Query`s are using `.onAppear` so only visible cells have active observations.

Again, neither `.always` nor `.onRender` were suitable for my original use case because they *don't* remain idle until the tab is selected for the first time. Until then, the active value observation noticeably slowed down my app.

---

`.always` and `.onAppear` behave as expected. I still don't think `.onRender` is useful, either I want to spare resources and will use `.onAppear` or I don't care and will use `.always`.
I think `.onRender` should either prefetch and then wait for `onAppear()` before live observing, or not exist at all.

Reply:
> `.always` and `.onAppear` behave as expected. I still don't think `.onRender` is useful, either I want to spare resources and will use `.onAppear` or I don't care and will use `.always`. I think `.onRender` should either prefetch and then wait for `onAppear()` before live observing, or not exist at all.

Thank you very, very much for your feedback :-)

Reply:
@kabouzeid I have significantly updated the documentation, based on our conversation. You can read it in the PR comment at the top.